clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -12414,6 +12414,14 @@ def err_sycl_kernel_incorrectly_named : Error<
   "'-fsycl-unnamed-lambda' to enable unnamed kernel lambdas"
   "}0">;
 
+// SYCL free function kernels extension.
+def err_bad_free_function_kernel_param_type : Error<
+  "%0 cannot be used as the type of a free function kernel parameter">;
+def note_free_function_kernel_param_type_not_fwd_declarable : Note<
+  "%0 is not forward declarable">;
+def note_free_function_kernel_param_type_not_supported : Note<
+  "%0 is not yet supported as free function kernel parameter">;
```

Comment:
I think having different diagnostics for free function parameter type and kernel param type is unnecessary. Other than the diagnostics you added/modified in this PR, won't other invalid types in free functions generate the old diagnostic `err_bad_kernel_param_type`? IMO I think we can keep the old diagnostic, passing type to it. The note diagnostic can be generated additionally for extra information

Reply:
I agree the diagnostics could be unified, but that would require changes to the existing diagnostics since their message text explicitly references the concept of a "kernel name". I'm not sure such unification would provide much benefit.
```
def err_nullptr_t_type_in_sycl_kernel : Error<"%0 is an invalid kernel name, "
                                        "'std::nullptr_t' is declared in the 'std' namespace ">;
def err_invalid_std_type_in_sycl_kernel : Error<"%0 is an invalid kernel name, " 
                                          "%q1 is declared in the 'std' namespace ">;
def err_sycl_kernel_incorrectly_named : Error<
  "%select{%1 is invalid; kernel name should be forward declarable "
  "at namespace scope"
  "|unscoped enum %1 requires fixed underlying type"
  "|unnamed type %1 is invalid; provide a kernel name, or use "
  "'-fsycl-unnamed-lambda' to enable unnamed kernel lambdas"
  "}0">;
```

Reply:
I used `err_bad_kernel_param_type` instead of a new one.

@tahonermann , I think @elizabethandrews meant another set of diagnostics.